### DISCUSSION TITLE
Be able to spectate a player, join/spectate their multiplayer, invite to a multiplayer game from anywhere.

### DISCUSSION THREAD

**Discussion by LeNitrous** [ppy/osu#14273] [Ideas]
As the title implies, the client should have the ability to spectate a player or join/spectate their multiplayer game, and invite them to your multiplayer game. Possible places may be:
- from chat by opening the context menu from their name: 
![image](https://user-images.githubusercontent.com/20495991/129359633-b50ebd13-492c-48f9-a4da-e6c728635678.png)

- opening a context menu from their user card:
![image](https://user-images.githubusercontent.com/20495991/129359764-314ca568-46bd-441d-a9dd-f66d9869efb0.png)

- from the user profile overlay (may also include web version):
![image](https://user-images.githubusercontent.com/20495991/129359855-5162e270-2571-4bde-accf-5217f82a0a20.png)

There is already an issue (https://github.com/ppy/osu/issues/8492) regarding joining via Discord's Rich Presence hence omitted from the main topic.

**Comment by peppy** [comment]
Adding spectate to these options should already be feasible. The others require further implementation in the underlying systems.

**Comment by minetoblend** [comment]
I'm interested in adding player invites for multiplayer and want to know if this is a contribution that you'd currently accept. If so, should this work just as in stable by sending a `osump://<roomId>/[password]` link through chat, or are there other plans for this?
In stable the messages also currently display the room name instead of the url, doing the same here would require loading the room info when formatting the link.

**Reply by peppy** [reply]
I think it's a good starting point, and yes, it would need to include the room details.

Of note though, on stable it is bancho that sends the invite since the client may not know the current password. You may find the same thing becomes an issue here as well (what happens if the host changes the password, for instance).

**Reply by minetoblend** [reply]
Since the client will have plain text access to the password through the link anways, is there anything speaking against the client just retrieving the password & sending the message from there instead of bancho? I could also see using some kinda token instead of the password working to avoid sending the plain text password but I think that might be overkill.

To retrieve the room info for formatting the link, would the best approach be to have some kind of async functionality to `MessageFormatter`? Perhaps including it in the invite link somehow would make things simpler.
Alternatively I could see adding support for async loading the text in `LinkText` as a workaround too.

**Reply by minetoblend** [reply]
Now that I think about it, would it be okay if I tried to do this using in-game notifications instead of chat messages? I feel like that would work out for better UX than chat messages and it would also make it easier to implement than somehow squeezing the room loading into the chat link formatting somehow & also easier to add localisation.
I was thinking of letting `osu-server-spectator` handle this, would keep things simple & would put this in the same place as the other multiplayer stuff. Drawback would be that the invites only work for people that are online this way, though I think that it doesn't make too much sense to invite people that are offline anyways. 
I'd still include osump:// link handling either way to support things like discord invites down the road

**Reply by peppy** [reply]
I think this sounds fine, yes. The only thing you'll need to consider with notifications is that they are bypassing the checks in place for chat, such as whether the user allows unsolicited PMs, and whether they have the other user blocked.

Don't necessarily have to add these back immediately, but should be TODO'd at a minimum.
